File: WINGS_AI_Bootcamp/Elixir/IntToRoman.ex
defmodule Roman do
  @val [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
  @sym ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"]

  def int_to_roman(num), do: int_to_roman(num, @val, @sym, "")

  defp int_to_roman(0, _val, _sym, acc), do: acc

  defp int_to_roman(num, [v | vt], [s | st], acc) when num >= v do
    int_to_roman(num - v, [v | vt], [s | st], acc <> s)
  end

  defp int_to_roman(num, [_v | vt], [_s | st], acc) do
    int_to_roman(num, vt, st, acc)
  end
end

IO.puts Roman.int_to_roman(1994)  # Output: MCMXCIV

File: WINGS_AI_Bootcamp/Haskell/IntToRoman.hs
import Data.List (zip)

intToRoman :: Int -> String
intToRoman num = go num valuesSymbols
  where
    valuesSymbols = [(1000,"M"),(900,"CM"),(500,"D"),(400,"CD"),
                     (100,"C"),(90,"XC"),(50,"L"),(40,"XL"),
                     (10,"X"),(9,"IX"),(5,"V"),(4,"IV"),(1,"I")]

    go 0 _ = ""
    go n ((val, sym):xs)
      | n >= val  = sym ++ go (n - val) ((val, sym):xs)
      | otherwise = go n xs

main :: IO ()
main = print $ intToRoman 1994  -- Output: "MCMXCIV"

